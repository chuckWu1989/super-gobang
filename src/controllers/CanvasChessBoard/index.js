const drawLine = (ctx, s, e) => {
  ctx.beginPath();
  ctx.moveTo(s[0], s[1]);
  ctx.lineTo(e[0], e[1]);
  ctx.closePath();
  ctx.stroke();
};
export const drawBoard = (canvas, props) => {
  const {
    grid,
    coord,
    width,
    ratio,
  } = props;
  const ctx = canvas.getContext('2d');
  const height = width * ratio;
  ctx.clearRect(0, 0, width, height);
  for (let i = 0; i <= (grid + 1); i += 1) {
    drawLine(ctx, coord[0][i], coord[grid + 1][i]);
    drawLine(ctx, coord[i][0], coord[i][grid + 1]);
  }
};

export default drawBoard;
