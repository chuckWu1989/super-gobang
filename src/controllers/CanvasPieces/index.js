import { radiusRatio, s, e } from './config';

const clearCanvas = (ctx, width, height) => {
  ctx.clearRect(0, 0, width, height);
};
const clearPiece = (ctx, coord, unitWidth) => {
  const cx = coord[0];
  const cy = coord[1];
  ctx.clearRect(cx - (unitWidth / 2), cy - (unitWidth / 2), unitWidth, unitWidth);
};
const drawPiece = (ctx, coord, unitWidth, color) => {
  const r = radiusRatio * (unitWidth / 2);
  const cx = coord[0];
  const cy = coord[1];
  ctx.beginPath();
  ctx.fillStyle = color;
  ctx.arc(cx, cy, r, s, e);
  ctx.fill();
  ctx.closePath();
};
const addOrRemovePiece = (ctx, nextProps, props) => {
  const { chessMoves: nextChessMoves, unitWidth, coord } = nextProps;
  const { chessMoves } = props;
  if (chessMoves.length < nextChessMoves.length) {
    const { x, y, color } = nextChessMoves[nextChessMoves.length - 1];
    const point = coord[x][y];
    drawPiece(ctx, point, unitWidth, color);
  } else if (chessMoves.length > nextChessMoves.length) {
    const { x, y } = chessMoves[chessMoves.length - 1];
    const point = coord[x][y];
    clearPiece(ctx, point, unitWidth);
  }
};
const redrawPieces = (ctx, props) => {
  const {
    width,
    height,
    unitWidth,
    coord,
  } = props;
  const { chessMoves } = props;
  clearCanvas(ctx, width, height);
  chessMoves.forEach((element) => {
    const { x, y, color } = element;
    const point = coord[x][y];
    drawPiece(ctx, point, unitWidth, color);
  });
};
export const detectChessMovesChange = (container, nextProps, props) => {
  const ctx = container.getContext('2d');
  const { chessMoves } = props;
  const { chessMoves: nextChessMoves } = nextProps;
  if (chessMoves.length !== nextChessMoves.length) {
    addOrRemovePiece(ctx, nextProps, props);
  }
};
export const detectWidthChange = (container, prevProps, props) => {
  const ctx = container.getContext('2d');
  const { width } = props;
  const { width: nextWidth } = prevProps;
  if (nextWidth !== width) {
    redrawPieces(ctx, prevProps);
  }
};

export default detectChessMovesChange;
