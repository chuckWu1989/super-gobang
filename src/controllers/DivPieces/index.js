import React from 'react';
import DivPiece from '../../views/DivPiece';
import { radiusRatio } from './config';

/**
 * Clear all content on chessboard
 * @param {array} children The children of component
 * @param {number} s Start point to be sliced
 * @param {number} e End point to be sliced
 * @returns {array} Updated children
 */
export function clearPiece(children, s, e) {
  const nextChildren = children.slice(s, e);
  return nextChildren;
}
/**
 * Draw piece on chessboard
 * @param {array} children The children of component
 * @param {array} coord The coordinate of piece
 * @param {number} unitWidth Unit width per grid
 * @param {string} color The color of piece
 * @returns {array} Updated children
 */
export const drawPiece = (children, coord, unitWidth, color) => {
  const size = radiusRatio * unitWidth;
  const x = (coord[0] - (size / 2)) + 1;
  const y = (coord[1] - (size / 2)) + 1;
  const key = `${x}-${y}-piece`;
  const piece = <DivPiece key={key} x={x} y={y} color={color} size={size} />;
  children.push(piece);
  return children;
};
/**
 * Rerender pieces on chessboard
 * @param {*} props The properties of the component
 */
export function redrawPieces(props) {
  const { unitWidth, coord, chessMoves } = props;
  const children = [];
  chessMoves.forEach((element) => {
    const { x, y, color } = element;
    const point = coord[x][y];
    drawPiece(children, point, unitWidth, color);
  });
  this.setState({ children });
}
/**
 * Add or remove piece on chessboard
 * @param {*} nextProps Next properties of the component
 * @param {*} props Current properties of the component
 */
export function addOrRemovePiece(nextProps, props) {
  const { chessMoves: nextChessMoves, unitWidth, coord } = nextProps;
  const { chessMoves } = props;
  const { children } = this.state;
  let nextChildren;
  if (chessMoves.length < nextChessMoves.length) {
    const { x, y, color } = nextChessMoves[nextChessMoves.length - 1];
    const point = coord[x][y];
    nextChildren = drawPiece(children, point, unitWidth, color);
  } else if (chessMoves.length > nextChessMoves.length) {
    const e = nextChessMoves.length - chessMoves.length;
    nextChildren = clearPiece(children, 0, e);
  }
  this.setState({ children: nextChildren.slice() });
}
/**
 * Detect if the chess moves
 * @param {object} nextProps Next properties of the component
 * @param {object} props Current properties of the component
 */
export function detectChessMovesChange(nextProps, props) {
  const { chessMoves: nextChessMoves } = nextProps;
  const { chessMoves } = props;
  if (nextChessMoves.length !== chessMoves.length) {
    addOrRemovePiece.call(this, nextProps, props);
  }
}
/**
 * Detect if the width of container is changed
 * @param {object} prevProps The previous properties of the component
 * @param {object} props Current properties of the component
 */
export function detectWidthChange(prevProps, props) {
  const { width } = props;
  const { width: prevWidth } = prevProps;
  if (prevWidth !== width) {
    redrawPieces.call(this, props);
  }
}

export default detectChessMovesChange;
