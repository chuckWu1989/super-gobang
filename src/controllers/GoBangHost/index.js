import { chainNum, black, white, deadlock } from './config';

/**
 * Check if the point of piece is in boundary of grid
 * @param {number} x The x coordinate of piece
 * @param {number} y The y coordinate of piece
 * @param {number} grid The grid number of gobang
 * @returns {boolean} Checked result
 */
export const isInBoundOfBoard = (x, y, grid) => x > 0 && x <= grid && y > 0 && y <= grid;
/**
 * Update chess moves
 * @param {array} chessMoves Array of the recordings of chess moves
 * @param {number} x The x coordinate of piece
 * @param {number} y The y coordinate of piece
 * @param {object} props The properties of the component
 * @returns {object} Updated chess moves array, current player name and its color
 */
export const updateChessMoves = (chessMoves, x, y, props) => {
  const { wColor, bColor } = props;
  const nextChessMoves = chessMoves.slice();
  const color = (nextChessMoves.length + 1) % 2 === 0 ? wColor : bColor;
  const player = color === bColor ? black : white;
  nextChessMoves.push({ x, y, color });
  return { nextChessMoves, player, color };
};
/**
 * Translate screen XY coordinates into chessboard XY coordinates
 * @param {element} target The target element of chess pieces
 * @param {number} posX The X screen coordinates
 * @param {number} posY The Y screen coordinates
 * @param {number} grid The grid number of gobang
 * @returns {object} The chessboard XY coordinates
 */
export const calculateXY = (target, posX, posY, grid) => {
  const width = target.getAttribute('width');
  const height = target.getAttribute('height');
  const x = Math.floor(Math.round((posX / width) * (grid + 1)));
  const y = Math.floor(Math.round((posY / height) * (grid + 1)));
  return { x, y };
};
/**
 * Check if the point exists in recordings of chess moves
 * @param {number} x The x coordinate of piece
 * @param {number} y The y coordinate of piece
 * @param {array} chessMoves Array of the recordings of chess moves
 * @param {function} otherConditions Other conditions which want to be added in judgement
 * @returns {boolean} The checking result
 */
export const pointExisted = (x, y, chessMoves, otherConditions = () => true) => {
  const point = chessMoves.find(item => item.x === x && item.y === y && otherConditions(item));
  return point !== undefined;
};
/**
 * Calculate all win conditions
 * @param {number} x The x coordinate of piece
 * @param {number} y The y coordinate of piece
 * @param {number} grid The grid number of gobang
 * @returns {array} All conditions the player could win
 */
export const getWinConditions = (x, y, grid) => {
  const yLowerBound = (y - chainNum) + 1;
  const yUpperBound = (y + chainNum) - 1;
  const xLowerBound = (x - chainNum) + 1;
  const xUPperBound = (x + chainNum) - 1;
  const T = yLowerBound > 0 ?
    Array.from({ length: chainNum }, (item, idx) => [x, y - idx]) :
    undefined;
  const TR = xUPperBound <= grid && yLowerBound > 0 ?
    Array.from({ length: chainNum }, (item, idx) => [x + idx, y - idx]) :
    undefined;
  const R = xUPperBound <= grid ?
    Array.from({ length: chainNum }, (item, idx) => [x + idx, y]) :
    undefined;
  const RB = xUPperBound <= grid && yUpperBound <= grid ?
    Array.from({ length: chainNum }, (item, idx) => [x + idx, y + idx]) :
    undefined;
  const B = yUpperBound <= grid ?
    Array.from({ length: chainNum }, (item, idx) => [x, y + idx]) :
    undefined;
  const LB = xLowerBound > 0 && yUpperBound <= grid ?
    Array.from({ length: chainNum }, (item, idx) => [x - idx, y + idx]) :
    undefined;
  const L = xLowerBound > 0 ?
    Array.from({ length: chainNum }, (item, idx) => [x - idx, y]) :
    undefined;
  const LT = xLowerBound > 0 && yLowerBound > 0 ?
    Array.from({ length: chainNum }, (item, idx) => [x - idx, y - idx]) :
    undefined;
  let result = [T, TR, R, RB, B, LB, L, LT];
  result = result.filter(item => item !== undefined);
  return result;
};
/**
 * Check if the player wins
 * @param {number} x The x coordinate of piece
 * @param {number} y The y coordinate of piece
 * @param {string} color The color of current piece
 * @param {number} grid The grid number of gobang
 * @param {array} chessMoves Array of the recordings of chess moves
 * @returns {boolean} The result if the player wins
 */
export function checkWin(x, y, color, grid, chessMoves) {
  if (chessMoves.length < 9) {
    return false;
  }
  const conditions = getWinConditions(x, y, grid);
  const result = conditions.some(condition => (
    condition.every(point => (
      pointExisted(point[0], point[1], chessMoves, item => item.color === color)
    ))
  ));
  return result;
}
/**
 * Whether game is over
 * @param {array} chessMoves Array of the recordings of chess moves
 * @param {number} x The x coordinate of piece
 * @param {number} y The y coordinate of piece
 * @param {string} color The color of current piece
 * @param {string} player The current player
 * @param {object} props The properties of the component
 * @returns {boolean} If true, the game is over.
 */
export const isGameOver = (chessMoves, x, y, color, player, props) => {
  const { gameOver, grid } = props;
  const result = checkWin(x, y, color, grid, chessMoves);
  if (result) {
    gameOver(player);
  } else if (chessMoves.length === grid * grid) {
    gameOver(deadlock);
  }
  return result;
};
/**
 * Update chess moves if point is valid
 * @param {*} x The x coordinate of piece
 * @param {*} y The y coordinate of piece
 * @param {*} chessMoves Array of the recordings of chess moves
 * @param {*} props The properties of the component
 */
export function updateIfPointIsValid(x, y, chessMoves, props) {
  const { grid } = props;
  if (isInBoundOfBoard(x, y, grid) && !pointExisted(x, y, chessMoves)) {
    const { nextChessMoves, player, color } = updateChessMoves(chessMoves, x, y, props);
    const nextLocked = isGameOver(nextChessMoves, x, y, color, player, props);
    this.setState({ chessMoves: nextChessMoves, locked: nextLocked });
  }
}
/**
 * Add piece on to the chessboard and check if the player wins or deadlocks
 * @param {*} props The properties of the component
 */
export function addPiece(props) {
  return (
    function listener(e) {
      const { grid } = props;
      const { target, offsetX, offsetY } = e;
      const isTarget = target.getAttribute('target');
      const { chessMoves, locked } = this.state;
      if (isTarget && !locked) {
        const { x, y } = calculateXY(target, offsetX, offsetY, grid);
        updateIfPointIsValid.call(this, x, y, chessMoves, props);
      }
    }
  );
}
/**
 * Fallback the piece
 * @param {*} props The properties of the component
 */
export function fallbackPieces(props) {
  const { singleRace } = props;
  const { chessMoves, locked } = this.state;
  const { length } = chessMoves;
  if (!locked) {
    let clonedChessMoves = chessMoves.slice();
    if (singleRace) {
      clonedChessMoves = clonedChessMoves.slice(0, length - 2);
    } else {
      clonedChessMoves = clonedChessMoves.slice(0, length - 1);
    }
    this.setState({ chessMoves: clonedChessMoves });
  }
}
/**
 * Restart the game
 */
export function restartGame() {
  this.setState({ chessMoves: [], locked: false });
}
/**
 * Add click event onto the chessboard
 * @param {*} props The propertis of the component
 */
export function addClickEventListener(props) {
  let listener = addPiece(props);
  listener = listener.bind(this);
  window.addEventListener('click', listener);
  this.setState({ listener });
}
/**
 * Remove click event from chessboard
 */
export function removeClickEventListener() {
  window.removeEventListener('click', this.state.listener);
}
/**
 * Check if fallback or restart is triggered
 * @param {*} nextProps The next properties of the component
 */
export function receivePropsHandler(nextProps) {
  const { fallback, restart } = nextProps;
  if (fallback === true) {
    fallbackPieces.call(this, nextProps);
  }
  if (restart === true) {
    restartGame.call(this);
  }
}
/**
 * Check if the component should be rerendered
 * @param {*} nextState The next state of the component
 * @param {*} state The current state of the component
 */
export const shouldUpdate = (nextState, state) => {
  const { chessMoves: nextChessMoves } = nextState;
  const { chessMoves } = state;
  return nextChessMoves !== chessMoves;
};

export default addClickEventListener;
