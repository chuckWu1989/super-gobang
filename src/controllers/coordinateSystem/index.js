export function setCoordinate() {
  const { width, ratio, grid } = this.props;
  if (typeof width !== 'number') {
    return false;
  }
  const coord = new Array(grid + 2);
  const trueWidth = Math.floor(width / 15) * 15;
  const trueHeight = Math.floor((trueWidth * ratio) / 15) * 15;
  const unitWidth = trueWidth / (grid + 1);
  const unitHeight = trueHeight / (grid + 1);
  for (let x = 0; x <= grid + 1; x += 1) {
    coord[x] = new Array(grid);
    for (let y = 0; y <= grid + 1; y += 1) {
      coord[x][y] = [unitWidth * x, unitHeight * y];
    }
  }
  return this.setState({ coord, width: trueWidth, height: trueHeight });
}

export default setCoordinate;
