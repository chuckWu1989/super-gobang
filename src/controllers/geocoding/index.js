/**
 * Calculate coordinates and set into state.
 */
export function setCoordinate() {
  const { width, ratio, grid } = this.props;
  if (typeof width !== 'number') {
    return false;
  }
  const coord = new Array(grid + 2);
  const unitWidth = Math.round(width / (grid + 1));
  const unitHeight = Math.round((width * ratio) / (grid + 1));
  const trueWidth = unitWidth * (grid + 1);
  const trueHeight = unitHeight * (grid + 1);
  for (let x = 0; x <= grid + 1; x += 1) {
    coord[x] = new Array(grid);
    for (let y = 0; y <= grid + 1; y += 1) {
      coord[x][y] = [unitWidth * x, unitHeight * y];
    }
  }
  return this.setState({
    coord,
    unitWidth,
    unitHeight,
    width: trueWidth,
    height: trueHeight,
  });
}
/**
 * Detect if the width of container is changed
 * @param {*} nextProps Next properties of the component
 * @param {*} props Current properties of the component
 */
export function detectWidthChanged(nextProps, props) {
  const { width: nextWidth } = nextProps;
  const { width } = props;
  if (nextWidth !== width) {
    setCoordinate.call(this);
  }
}

export default setCoordinate;
