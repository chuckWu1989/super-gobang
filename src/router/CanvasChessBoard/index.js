import { drawBoard } from '../../controllers/CanvasChessBoard';
import CanvasChessBoard from '../../views/CanvasChessBoard';
import { connect } from '../../hocs/connect';

export const method = props => ({
  drawBoard: chessboard => drawBoard(chessboard, props),
});

export default connect(method)(CanvasChessBoard);
