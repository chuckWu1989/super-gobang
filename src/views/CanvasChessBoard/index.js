import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { drawBoard } from '../../controllers/CanvasChessBoard';
import './style.less';

class CanvasChessBoard extends Component {
  static propTypes = {
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,
  }
  componentDidMount() {
    drawBoard(this.chessboard, this.props);
  }
  componentDidUpdate() {
    drawBoard(this.chessboard, this.props);
  }
  render() {
    const { width, height } = this.props;
    return (
      <canvas
        width={width}
        height={height}
        ref={(self) => { this.chessboard = self; }}
        className="canvas-chessboard-style"
      />
    );
  }
}

export default CanvasChessBoard;
