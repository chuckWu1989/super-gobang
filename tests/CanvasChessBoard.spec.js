import React from 'react';
import toJson from 'enzyme-to-json';
import CanvasChessBoard from '../src/views/CanvasChessBoard';
import { method } from '../src/router/CanvasChessBoard';
import * as controller from '../src/controllers/CanvasChessBoard';

const { shallow } = global;
const coord = Array.from({ length: 17 }, (v, i) => {
  const arr = [];
  for (let j = 0; j <= 16; j += 1) arr.push([i * 10, j * 10]);
  return arr;
});
const props = {
  width: 160,
  height: 160,
  ratio: 1,
  grid: 15,
  coord,
};

describe('CanvasChessBoard', () => {
  describe('view', () => {
    it('should be rendered properly', () => {
      const ownProps = { ...props };
      const tree = toJson(shallow(<CanvasChessBoard {...ownProps} />));
      expect(tree).toMatchSnapshot();
    });
    it('should call methods during lifecycle of component', () => {
      const ownProps = {
        ...props,
        drawBoard: jest.fn(),
      };
      const wrapper = shallow(
        <CanvasChessBoard {...ownProps} />,
        { lifecycleExperimental: true },
      );
      wrapper.setProps({ width: 500 });
      expect(ownProps.drawBoard).toHaveBeenCalledTimes(2);
    });
  });
  describe('router', () => {
    it('should pass drawBoard method', () => {
      expect('drawBoard' in method()).toBeTruthy();
    });
  });
  describe('controller', () => {
    let canvas;
    beforeEach(() => {
      canvas = document.createElement('canvas');
    });
    describe('drawLine', () => {
      it('should draw line properly by start point and end point', () => {
        const ctx = canvas.getContext('2d');
        const oriBase64String = canvas.toDataURL();
        const s = [100, 0];
        const e = [100, 100];
        controller.drawLine(ctx, s, e);
        const base64String = canvas.toDataURL();
        expect(base64String).not.toBe(oriBase64String);
      });
    });
    describe('drawBoard', () => {
      it('should draw board properly', () => {
        const oriBase64String = canvas.toDataURL();
        controller.drawBoard(canvas, props);
        const base64String = canvas.toDataURL();
        expect(base64String).not.toBe(oriBase64String);
      });
    });
  });
});
