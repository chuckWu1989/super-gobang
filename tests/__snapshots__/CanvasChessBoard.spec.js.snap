// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CanvasChessBoard view should be rendered properly 1`] = `
<canvas
  className="canvas-chessboard-style"
  height={160}
  width={160}
/>
`;
